과제(6/30)<br>

Q1 ~ Q10을 수행하세요

#### 그룹연산
- 복잡한 데이터를 어떤 기준에 따라 여러 그룹으로 나누어서 관찰할 수 있으며 이런 방식으로 분할 처리하는 것을 그룹 연산이라 함
- 그룹 연산은 데이터를 집계, 반환, 필터하는데 효율적이며 판다스 groupby()메소드를 사용
- 그룹 객체 만들기(분할)
- 그룹 연산 메소드(적용-결합)

In [33]:
import pandas as pd
import seaborn as sns

titanic = sns.load_dataset('titanic')
titanic.columns

#pclass와 class는 같은 것, 둘 중에 하나는 제거해야 함

Index(['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare',
       'embarked', 'class', 'who', 'adult_male', 'deck', 'embark_town',
       'alive', 'alone'],
      dtype='object')

In [3]:
import pandas as pd
import seaborn as sns

titanic = sns.load_dataset('titanic')
df = titanic.loc[:,['age','sex','class','fare','survived']]

print('승객수:', len(df))
print(df.head())

승객수: 891
    age     sex  class     fare  survived
0  22.0    male  Third   7.2500         0
1  38.0  female  First  71.2833         1
2  26.0  female  Third   7.9250         1
3  35.0  female  First  53.1000         1
4  35.0    male  Third   8.0500         0


In [4]:
print(df.columns)

Index(['age', 'sex', 'class', 'fare', 'survived'], dtype='object')


In [5]:
# class열을 기준으로 그룹핑(분할)
grouped = df.groupby(['class'])
print(grouped)
list(grouped)

[('First',
        age     sex  class     fare  survived
  1    38.0  female  First  71.2833         1
  3    35.0  female  First  53.1000         1
  6    54.0    male  First  51.8625         0
  11   58.0  female  First  26.5500         1
  23   28.0    male  First  35.5000         1
  ..    ...     ...    ...      ...       ...
  871  47.0  female  First  52.5542         1
  872  33.0    male  First   5.0000         0
  879  56.0  female  First  83.1583         1
  887  19.0  female  First  30.0000         1
  889  26.0    male  First  30.0000         1
  
  [216 rows x 5 columns]),
 ('Second',
        age     sex   class     fare  survived
  9    14.0  female  Second  30.0708         1
  15   55.0  female  Second  16.0000         1
  17    NaN    male  Second  13.0000         1
  20   35.0    male  Second  26.0000         0
  21   34.0    male  Second  13.0000         1
  ..    ...     ...     ...      ...       ...
  866  27.0  female  Second  13.8583         1
  874  28.0  female

In [6]:
# 그룹 객체를 iteration으로 출력
for key, group in grouped:
    print('* key :', key)
    print('* number :', len(group))
    print(group.head(), '\n')

* key : First
* number : 216
     age     sex  class     fare  survived
1   38.0  female  First  71.2833         1
3   35.0  female  First  53.1000         1
6   54.0    male  First  51.8625         0
11  58.0  female  First  26.5500         1
23  28.0    male  First  35.5000         1 

* key : Second
* number : 184
     age     sex   class     fare  survived
9   14.0  female  Second  30.0708         1
15  55.0  female  Second  16.0000         1
17   NaN    male  Second  13.0000         1
20  35.0    male  Second  26.0000         0
21  34.0    male  Second  13.0000         1 

* key : Third
* number : 491
    age     sex  class     fare  survived
0  22.0    male  Third   7.2500         0
2  26.0  female  Third   7.9250         1
4  35.0    male  Third   8.0500         0
5   NaN    male  Third   8.4583         0
7   2.0    male  Third  21.0750         0 



In [7]:
# 연산 메서드 적용
average = grouped.mean()#그룹 객체 평균 값 구하기
print(average)
#survived의 값은 생존율

              age       fare  survived
class                                 
First   38.233441  84.154687  0.629630
Second  29.877630  20.662183  0.472826
Third   25.140620  13.675550  0.242363


In [8]:
stat = grouped.max()
stat

age   sex      fare  survived
class                                 
First   80.0  male  512.3292         1
Second  70.0  male   73.5000         1
Third   74.0  male   69.5500         1

In [9]:
# Q. 3rd 그룹만을 선택해서 group3 이름으로 저장하고 통계요약표를 출력하세요.
group3 = grouped.get_group('Third')
group3.isnull().sum()

age         136
sex           0
class         0
fare          0
survived      0
dtype: int64

In [10]:
grouped_two = df.groupby(['class','sex'])
average_two = grouped_two.mean()
average_two

age        fare  survived
class  sex                                    
First  female  34.611765  106.125798  0.968085
       male    41.281386   67.226127  0.368852
Second female  28.722973   21.970121  0.921053
       male    30.740707   19.741782  0.157407
Third  female  21.750000   16.118810  0.500000
       male    26.507589   12.661633  0.135447

In [11]:
group3f = grouped_two.get_group(('Third','female'))
group3f.head()

age     sex  class     fare  survived
2   26.0  female  Third   7.9250         1
8   27.0  female  Third  11.1333         1
10   4.0  female  Third  16.7000         1
14  14.0  female  Third   7.8542         0
18  31.0  female  Third  18.0000         0

In [12]:
# 모든 열에 여러 함수를 매핑
agg_all = grouped.agg(['min','max'])
agg_all.head()

age           sex       fare           survived    
         min   max     min   max  min       max      min max
class                                                       
First   0.92  80.0  female  male  0.0  512.3292        0   1
Second  0.67  70.0  female  male  0.0   73.5000        0   1
Third   0.42  74.0  female  male  0.0   69.5500        0   1

In [13]:
# 각 열마다 다른 함수를 적용하여 집계
agg_sep = grouped.agg({'fare':['min','max'], 'age':'mean'})
agg_sep.head()

fare                  age
        min       max       mean
class                           
First   0.0  512.3292  38.233441
Second  0.0   73.5000  29.877630
Third   0.0   69.5500  25.140620

In [14]:
# Q. 데이터 개수가 200개 이상인 그룹만을 필터링하여 데이터프레임으로 반환하여 grouped_filter 이름으로
#저장 후 처음 5개 행을 출력하세요.

grouped_filter = grouped.filter(lambda x: len(x) >= 200)
grouped_filter.head()


age     sex  class     fare  survived
0  22.0    male  Third   7.2500         0
1  38.0  female  First  71.2833         1
2  26.0  female  Third   7.9250         1
3  35.0  female  First  53.1000         1
4  35.0    male  Third   8.0500         0

In [15]:
grouped_filter['class'].value_counts()

Third     491
First     216
Second      0
Name: class, dtype: int64

### 과제(7/1). 
-age열의 평균이 30보다 작은 그룹만을 필터링하여 데이터프레임으로 반환하여 age_filter 이름으로 저장 후 처음 5개 행을 출력하세요.

In [16]:
age_filter= grouped.filter(lambda x:x.agg({'age': 'mean'})<30)
age_filter.head()

age     sex  class     fare  survived
0  22.0    male  Third   7.2500         0
2  26.0  female  Third   7.9250         1
4  35.0    male  Third   8.0500         0
5   NaN    male  Third   8.4583         0
7   2.0    male  Third  21.0750         0

In [17]:
age_filter['class'].value.counts()

AttributeError: 'Series' object has no attribute 'value'

## pivot

피벗 테이블을 구성하는 4가지 요소(행인덱스, 열인덱스, 데이터 값, 데이터 집계함수)에 적용할 데이터프레임의 열을 각각 지정하여 함수의 인자로 전달

In [ ]:
import seaborn as sns
titanic = sns.load_dataset('titanic')
df = titanic.loc[:,['age','sex','class','fare','survived']]
df.head()

In [136]:
pdf1 = pd.pivot_table(df,
                     index = 'class',
                     columns = 'sex',
                     values = 'age',
                     aggfunc = 'mean')
pdf1.head()

sex        female       male
class                       
First   34.611765  41.281386
Second  28.722973  30.740707
Third   21.750000  26.507589

In [137]:
# 행, 열, 값에 사용할 열을 2개이상 지정 - index, value, aggfunc에 각각 2개 이상 지정
pdf2 = pd.pivot_table(df,
                     index = ['class','sex'],
                     columns='survived',
                     values=['age','fare'],
                     aggfunc=['mean','max'])
pdf2

mean                                      max        \
                     age                   fare               age         
survived               0          1           0           1     0     1   
class  sex                                                                
First  female  25.666667  34.939024  110.604167  105.978159  50.0  63.0   
       male    44.581967  36.248000   62.894910   74.637320  71.0  80.0   
Second female  36.000000  28.080882   18.250000   22.288989  57.0  55.0   
       male    33.369048  16.022000   19.488965   21.095100  70.0  62.0   
Third  female  23.818182  19.329787   19.773093   12.464526  48.0  63.0   
       male    27.255814  22.274211   12.204469   15.579696  74.0  45.0   

                                 
                 fare            
survived            0         1  
class  sex                       
First  female  151.55  512.3292  
       male    263.00  512.3292  
Second female   26.00   65.0000  
       male     73.50   39.0000  
Third  female   69.55   31.3875  
       male     69.55   56.4958

In [138]:
#결측값
df_t = sns.load_dataset('titanic')
df_t.isnull().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

In [139]:
df_t.deck.value_counts()

C    59
B    47
D    33
E    32
A    15
F    13
G     4
Name: deck, dtype: int64

In [107]:
df_t.embarked.value_counts()

S    644
C    168
Q     77
Name: embarked, dtype: int64

In [108]:
#누락데이터의 앞의 값으로 대체(ffill함수는 앞의 값으로 대체, bfill은 데이터의 뒤의 값으로 대체된다.)
df_t.embarked.fillna(method='ffill',inplace=True)
df_t.embarked.value_counts()

S    644
C    169
Q     78
Name: embarked, dtype: int64

### 과제(7/1)
- age 컬럼의 결측치를 적절한 값으로 대체하시고 그 근거를 설명하세요.
- deck 컬럼의 삭제 여부를 판단하는 근거를 설명하세요.

In [212]:
#1) age 컬럼
import numpy as np
df_t.age.describe()

count    714.000000
mean      29.699118
std       14.526497
min        0.420000
25%       20.125000
50%       28.000000
75%       38.000000
max       80.000000
Name: age, dtype: float64

In [195]:
df_t.age.info

<bound method Series.info of 0      22.0
1      38.0
2      26.0
3      35.0
4      35.0
       ... 
886    27.0
887    19.0
888     NaN
889    26.0
890    32.0
Name: age, Length: 891, dtype: float64>

In [213]:
df_t.age.value_counts()

24.00    30
22.00    27
18.00    26
19.00    25
28.00    25
         ..
36.50     1
55.50     1
0.92      1
23.50     1
74.00     1
Name: age, Length: 88, dtype: int64

In [216]:
df_t.age.isnull().sum()

177

In [217]:
177/891

0.19865319865319866

In [ ]:
#'age'컬럼의 최빈값은 24살, 최빈 나이대는 10대 후반에서 20대이다. 평균값은 29살, 백분위수 50%,  28살, 50%는 중앙값하고 비슷하다.
#또한 결측치의 전체에서의 비중은 177/891 = 약 20%정도 먼저 평균값으로 대체해보겠다.

In [231]:
df_t.age.replace(np.nan, "28",inplace=True)

In [232]:
df_t.age.isnull().sum()

0

In [233]:
df_t.age.value_counts()

29.699118    177
24.000000     30
22.000000     27
18.000000     26
28.000000     25
            ... 
36.500000      1
55.500000      1
0.920000       1
23.500000      1
74.000000      1
Name: age, Length: 89, dtype: int64

In [234]:
df_t.age.describe()

count    891.000000
mean      29.699118
std       13.002015
min        0.420000
25%       22.000000
50%       29.699118
75%       35.000000
max       80.000000
Name: age, dtype: float64

In [ ]:
#age열의 NaN값을 평균으로 대치되었을 때, 평균값은 변화가 없고, 표준편차는 한 값의 양이 많아져서 
#조금 줄어든 것을 알 수 있다. 또한 중앙값이 최빈값인 29.6세로 바뀌었다. 
#전체적 통계값 중에 중앙값과 최빈값, 표준편차가 바뀌었다,
#전체적 통계값에 오염이 있다면은 최빈값과 중앙값, 평균이 같아졌다는 점이다. 
#하지만 전체적 통계값의 의미에 큰 영향을 주었는 가에 대해서는
#표준편차가 조금 작아졌고 표준편차의 의미에 의하면 좀 더 집단을 대표하는 통계값이 되었다고 할 수 있다.
#그러므로 age열의 결측치를 평균으로 바꾸겠다.

In [ ]:
# 2) deck컬럼

In [237]:
df_t.deck.value_counts()

C    59
B    47
D    33
E    32
A    15
F    13
G     4
Name: deck, dtype: int64

In [239]:
df_t.deck.isnull().sum()

688

In [240]:
df_t.deck.info

<bound method Series.info of 0      NaN
1        C
2      NaN
3        C
4      NaN
      ... 
886    NaN
887      B
888    NaN
889      C
890    NaN
Name: deck, Length: 891, dtype: category
Categories (7, object): ['A', 'B', 'C', 'D', 'E', 'F', 'G']>

In [241]:
df_t.deck.describe()

count     203
unique      7
top         C
freq       59
Name: deck, dtype: object

In [ ]:
#deck컬럼은 'A', 'B', 'C', 'D', 'E', 'F', 'G'와 같은 알파벳으로 이뤄져 있다. 
#그리고 deck의 사전적 의미는 배의 갑판을 의미한다. 그렇다면 어떠한 승객들의 배 안에서의 위치를 나타낼지도 모른다.
#그렇다면 저 알파벳이 특별한 의미가 있는지에 대해 알아보기 위해서 ,선실의 'class'와 결부시켜보겠다. 값은 'fare'이고 임의로 평균값을 썼다.
#그 다음엔 surived의 평균값으로 각 클래스와 deck별 생존율을 알아봤다.
#deck의 의미는 덱 위치별 생존자들이 찾아진 정도 인 것과 같다. 많이 살아남은 퍼스트 클래스가 당연히 많은 생존율을 보이고
#살아남지 못한 써드는 큰 의미 보여주지 못한다.
# 이 값의 NaN 값은 찾아지지 않았거나 덱의 알파벳을 받지 못한 사람들일 가능성이 높으므로 필요 없는 수치이다.
#그래서 이 값을 제거하겠다.

In [247]:
pdf1 = pd.pivot_table(df_t,
                     index = 'deck',
                     columns = 'class',
                     values = 'survived',
                     aggfunc = 'mean')
pdf1.head()

class     First  Second  Third
deck                          
A      0.466667     NaN    NaN
B      0.744681     NaN    NaN
C      0.593220     NaN    NaN
D      0.758621    0.75    NaN
E      0.720000    0.75    1.0

In [248]:
df_t.dropna(subset=['deck'], inplace=True)

In [249]:
df_t.deck.isnull().sum()

0

In [250]:
df_t.deck.info

<bound method Series.info of 1      C
3      C
6      E
10     G
11     C
      ..
871    D
872    B
879    C
887    B
889    C
Name: deck, Length: 203, dtype: category
Categories (7, object): ['A', 'B', 'C', 'D', 'E', 'F', 'G']>

```
df.dropna(subset=['column'], inplace=True) : 특정 컬럼의 널 값을 없앰
df.dropna(inplace = True) : 널 값을 없앰
fillna
df.column.replace(np.nan, df.column.mean(),inplace=True) : 널 값을 다른 걸로 대체
```

<img src='dataset/apply_applymap_pipe.jpg' STYLE='width:800px;'>
[pipe vs applymap vs apply]  

pipe와 applymap과의 차이는 반환하는 결과가 pipe가 더 다양하다는 것으로 applymap은 데이터프레임 원소에 적용하므로 데이터프레임으로만 리턴되지만 pipe는 사용자 함수의 리턴값에 따라 개별 값, 시리즈, 데이터프레임을 반환할 수 있다.  

현재 가장 많이 쓰이는 메서드는 apply인데 그 이유는 다양한 형태로 값을 입력 받을 수 있고, 반환하는 형태도 폭이 넓기 때문. 결과는 같아보이지만, 시리즈를 이용해 데이터프레임의 행과 열도 이용할 수 있으며 시리즈 그 자체도 함수로 활용할 수 있다. 

만약 데이터프레임에 함수를 적용해야한다면 리턴하는 형태가 많은 pipe 사용도 고려


In [134]:
import pandas as pd
df = pd.DataFrame({
    'Col1':[30,40,50,60],
    'Col2':[23,35,65,45],
    'Col3':[85,87,90,89],
},index=list('ABCD'))
df 

Col1  Col2  Col3
A    30    23    85
B    40    35    87
C    50    65    90
D    60    45    89

In [135]:
df.apply(lambda x:x.sum(),axis=0)

Col1    180
Col2    168
Col3    351
dtype: int64

In [136]:
df.Col1.apply(lambda x:x+1)

A    31
B    41
C    51
D    61
Name: Col1, dtype: int64

In [137]:
df.Col1.map(lambda x:x+1)

A    31
B    41
C    51
D    61
Name: Col1, dtype: int64

In [138]:
df.applymap(lambda x:"%.2f"% x)

Col1   Col2   Col3
A  30.00  23.00  85.00
B  40.00  35.00  87.00
C  50.00  65.00  90.00
D  60.00  45.00  89.00

In [139]:
# The pipe() method allows you to apply one or more functions to the DataFrame object.

import pandas as pd

def change_age(x):
  x["age"]=[10, 20, 30]
  return x

data = {
  "name": ["Sally", "Mary", "John"],
  "age": [50, 40, 30]
}

df = pd.DataFrame(data)
df

name  age
0  Sally   50
1   Mary   40
2   John   30

In [140]:
df.pipe(change_age)

name  age
0  Sally   10
1   Mary   20
2   John   30

### 과제
- Q11~Q15 문제 수행

In [141]:
#중복
df = pd.DataFrame({'c1':['a','a','b','a','b'],
                  'c2':[1,1,1,2,2],
                  'c3':[1,1,2,2,2]})
df

c1  c2  c3
0  a   1   1
1  a   1   1
2  b   1   2
3  a   2   2
4  b   2   2

In [143]:
#전체 행 데이터 중에서 중복 확인
df_dup=df.duplicated()
df_dup

0    False
1     True
2    False
3    False
4    False
dtype: bool

In [144]:
# 특정 열 데이터 중에서 중복
col_dup = df.c2.duplicated()
col_dup

0    False
1     True
2     True
3    False
4     True
Name: c2, dtype: bool

In [146]:
# drop_duplicates() : 중복 행 제거
df2 = df.drop_duplicates()
df2

c1  c2  c3
0  a   1   1
2  b   1   2
3  a   2   2
4  b   2   2

In [147]:
# 특정 컬럼 중복 제거
df3 = df.drop_duplicates(subset=['c2','c3'])
df3

c1  c2  c3
0  a   1   1
2  b   1   2
3  a   2   2

In [288]:
# 데이터 단위 변경
auto_df = pd.read_csv('dataset/auto-mpg.csv')
auto_df.head()

mpg  cylinders  displacement horsepower  weight  acceleration  model year  \
0  18.0          8         307.0        130    3504          12.0          70   
1  15.0          8         350.0        165    3693          11.5          70   
2  18.0          8         318.0        150    3436          11.0          70   
3  16.0          8         304.0        150    3433          12.0          70   
4  17.0          8         302.0        140    3449          10.5          70   

   origin                   car name  
0       1  chevrolet chevelle malibu  
1       1          buick skylark 320  
2       1         plymouth satellite  
3       1              amc rebel sst  
4       1                ford torino

In [289]:
# mpg를 kpl로 환산
mpg_to_kpl = 1.60934 / 3.78541
mpg_to_kpl

0.42514285110463595

In [290]:
auto_df['kpl'] = auto_df['mpg'].apply(lambda x:x*mpg_to_kpl)
auto_df[['mpg','kpl']].head()

mpg       kpl
0  18.0  7.652571
1  15.0  6.377143
2  18.0  7.652571
3  16.0  6.802286
4  17.0  7.227428

In [291]:
auto_df.horsepower.unique()

array(['130', '165', '150', '140', '198', '220', '215', '225', '190',
       '170', '160', '95', '97', '85', '88', '46', '87', '90', '113',
       '200', '210', '193', '?', '100', '105', '175', '153', '180', '110',
       '72', '86', '70', '76', '65', '69', '60', '80', '54', '208', '155',
       '112', '92', '145', '137', '158', '167', '94', '107', '230', '49',
       '75', '91', '122', '67', '83', '78', '52', '61', '93', '148',
       '129', '96', '71', '98', '115', '53', '81', '79', '120', '152',
       '102', '108', '68', '58', '149', '89', '63', '48', '66', '139',
       '103', '125', '133', '138', '135', '142', '77', '62', '132', '84',
       '64', '74', '116', '82'], dtype=object)

In [292]:
import numpy as np
auto_df.horsepower.replace('?', np.nan, inplace = True)
auto_df.dropna(subset = ['horsepower'],axis=0,inplace = True)
auto_df.horsepower.isnull().sum()

0

In [293]:
auto_df.horsepower = auto_df.horsepower.astype('float')
auto_df.horsepower.dtypes

dtype('float64')

In [294]:
auto_df = pd.read_csv('dataset/auto-mpg.csv')
auto_df.head()

mpg  cylinders  displacement horsepower  weight  acceleration  model year  \
0  18.0          8         307.0        130    3504          12.0          70   
1  15.0          8         350.0        165    3693          11.5          70   
2  18.0          8         318.0        150    3436          11.0          70   
3  16.0          8         304.0        150    3433          12.0          70   
4  17.0          8         302.0        140    3449          10.5          70   

   origin                   car name  
0       1  chevrolet chevelle malibu  
1       1          buick skylark 320  
2       1         plymouth satellite  
3       1              amc rebel sst  
4       1                ford torino

### 과제(7/1)
- '?'가 있는 인덱스(불린 인덱싱)를 찾아서 그 행을 삭제하세요.

In [298]:
dff= auto_df[auto_df == "?"]
dff.head()

mpg  cylinders  displacement horsepower  weight  acceleration  model year  \
0  NaN        NaN           NaN        NaN     NaN           NaN         NaN   
1  NaN        NaN           NaN        NaN     NaN           NaN         NaN   
2  NaN        NaN           NaN        NaN     NaN           NaN         NaN   
3  NaN        NaN           NaN        NaN     NaN           NaN         NaN   
4  NaN        NaN           NaN        NaN     NaN           NaN         NaN   

   origin car name  
0     NaN      NaN  
1     NaN      NaN  
2     NaN      NaN  
3     NaN      NaN  
4     NaN      NaN

In [252]:
import numpy as np
auto_df.replace('?', np.nan, inplace = True) # ?를 찾기 위해서 데이터프레임의 ?를 NaN깂으로 바꿨다.
#horsepower의 6개의 ?값이 전체 데이터프레임의 ?값의 전부란 것을 알 수 있다.

In [253]:
auto_df.isnull().sum()

mpg             0
cylinders       0
displacement    0
horsepower      6
weight          0
acceleration    0
model year      0
origin          0
car name        0
dtype: int64

In [255]:
auto_df.dropna(inplace = True)#?가 NaN으로 바뀐 행을 삭제한다.

In [256]:
auto_df.isnull().sum() #NaN값이 없어진 것을 확인하면서 ?가 있는 행이 사라졌음을 알 수 있다.

mpg             0
cylinders       0
displacement    0
horsepower      0
weight          0
acceleration    0
model year      0
origin          0
car name        0
dtype: int64

## (7/4)

In [183]:
import pandas as pd
auto_df = pd.read_csv('dataset/auto-mpg.csv')
auto_df.head()

mpg  cylinders  displacement horsepower  weight  acceleration  model year  \
0  18.0          8         307.0        130    3504          12.0          70   
1  15.0          8         350.0        165    3693          11.5          70   
2  18.0          8         318.0        150    3436          11.0          70   
3  16.0          8         304.0        150    3433          12.0          70   
4  17.0          8         302.0        140    3449          10.5          70   

   origin                   car name  
0       1  chevrolet chevelle malibu  
1       1          buick skylark 320  
2       1         plymouth satellite  
3       1              amc rebel sst  
4       1                ford torino

In [184]:
auto_df.origin.unique()

array([1, 3, 2], dtype=int64)

Q. 아래 사항을 처리 하세요.
* origin 열의 고유값을 출력하세요.
* 정수형 데이터를 문자형 데이터로 변환한 후 고유값을 출력하세요.(1:'USA', 2:'EU', 3:'JAPAN')


In [187]:
oror = auto_df[["origin"]]
oror.head()

origin
0       1
1       1
2       1
3       1
4       1

In [202]:
auto_df.replace(1,"USA", inplace=True)
auto_df.replace(2,"EU",inplace=True)
auto_df.replace(3, "JAPAN",inplace=True)
auto_df.head()

mpg cylinders  displacement horsepower  weight  acceleration  model year  \
0  18.0         8         307.0        130    3504          12.0          70   
1  15.0         8         350.0        165    3693          11.5          70   
2  18.0         8         318.0        150    3436          11.0          70   
3  16.0         8         304.0        150    3433          12.0          70   
4  17.0         8         302.0        140    3449          10.5          70   

  origin                   car name  
0    USA  chevrolet chevelle malibu  
1    USA          buick skylark 320  
2    USA         plymouth satellite  
3    USA              amc rebel sst  
4    USA                ford torino

In [203]:
auto_df.origin = auto_df.origin.astype('category')
auto_df.origin.dtypes

CategoricalDtype(categories=['USA', 'EU', 'JAPAN'], ordered=False)

In [204]:
auto_df.origin = auto_df.origin.astype('str')
auto_df.origin.dtypes

dtype('O')

In [208]:
auto_df = pd.read_pickle('dataset/auto-mpg.pkl')
auto_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 392 entries, 0 to 397
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   mpg           392 non-null    float64
 1   cylinders     392 non-null    int64  
 2   displacement  392 non-null    float64
 3   horsepower    392 non-null    float64
 4   weight        392 non-null    int64  
 5   acceleration  392 non-null    float64
 6   model year    392 non-null    int64  
 7   origin        392 non-null    int64  
 8   car name      392 non-null    object 
dtypes: float64(4), int64(4), object(1)
memory usage: 30.6+ KB


In [211]:
#horsepower를 3개의 bin으로 나누고 bin_names = ['저출력','보통출력','고출력']으로 출력하세요.
import numpy as np
count, bin_dividers = np.histogram(auto_df.horsepower, bins = 3)
print(bin_dividers)#몇 구간으로 나누는 변수
bin_names = ['저출력','보통출력','고출력']
auto_df['hp_bin'] = pd.cut(x = auto_df.horsepower, #데이터 배열
                          bins=bin_dividers,#경계값 리스트, 구간으로 나눴다.
                          labels = bin_names,#bin 이름
                          include_lowest = True)#첫 경계값 포함
auto_df[['horsepower','hp_bin']].head()

[ 46.         107.33333333 168.66666667 230.        ]


horsepower hp_bin
0       130.0   보통출력
1       165.0   보통출력
2       150.0   보통출력
3       150.0   보통출력
4       140.0   보통출력

In [212]:
auto_df.horsepower.unique()

array([130., 165., 150., 140., 198., 220., 215., 225., 190., 170., 160.,
        95.,  97.,  85.,  88.,  46.,  87.,  90., 113., 200., 210., 193.,
       100., 105., 175., 153., 180., 110.,  72.,  86.,  70.,  76.,  65.,
        69.,  60.,  80.,  54., 208., 155., 112.,  92., 145., 137., 158.,
       167.,  94., 107., 230.,  49.,  75.,  91., 122.,  67.,  83.,  78.,
        52.,  61.,  93., 148., 129.,  96.,  71.,  98., 115.,  53.,  81.,
        79., 120., 152., 102., 108.,  68.,  58., 149.,  89.,  63.,  48.,
        66., 139., 103., 125., 133., 138., 135., 142.,  77.,  62., 132.,
        84.,  64.,  74., 116.,  82.])

In [ ]:
# Q. horsepower 열의 최대값의 절대값으로 모든 데이터를 나누는 정규화를 수행한 후 통계요약표를 출력하세요.


In [254]:

def maxdiv(x):
    return x/abs(max(x))

df2 = auto_df[["horsepower"]].apply(maxdiv)
df2.describe()

horsepower
count  392.000000
mean     0.454215
std      0.167353
min      0.200000
25%      0.326087
50%      0.406522
75%      0.547826
max      1.000000

In [3]:
import pandas as pd
fifa = pd.read_csv('dataset/FIFA.csv',encoding='cp949')
fifa.head()

ID               Name  Age Nationality  Overall                 Club  \
0  158023           L. Messi   31   Argentina       94         FC Barcelona   
1   20801  Cristiano Ronaldo   33    Portugal       94             Juventus   
2  190871          Neymar Jr   26      Brazil       92  Paris Saint-Germain   
3  193080             De Gea   27       Spain       91    Manchester United   
4  192985       K. De Bruyne   27     Belgium       91      Manchester City   

  Preferred_Foot       Work_Rate Position  Jersey_Number  \
0           Left  Medium/ Medium       RF             10   
1          Right       High/ Low       ST              7   
2          Right    High/ Medium       LW             10   
3          Right  Medium/ Medium       GK              1   
4          Right      High/ High      RCM              7   

   Contract_Valid_Until Height  Weight_lb  Release_Clause   Value  Wage  
0                  2021    5'7        159          226500  110500   565  
1                  2022    6'2        183          127100   77000   405  
2                  2022    5'9        150          228100  118500   290  
3                  2020    6'4        168          138600   72000   260  
4                  2023   5'11        154          196400  102000   355

### 과제(7/4) : height의 값을 cm로 변환하여 새로운 컬럼 'hight_cm'을 생성하세요.


In [5]:
fifa.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16642 entries, 0 to 16641
Data columns (total 16 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   ID                    16642 non-null  int64 
 1   Name                  16642 non-null  object
 2   Age                   16642 non-null  int64 
 3   Nationality           16642 non-null  object
 4   Overall               16642 non-null  int64 
 5   Club                  16642 non-null  object
 6   Preferred_Foot        16642 non-null  object
 7   Work_Rate             16642 non-null  object
 8   Position              16642 non-null  object
 9   Jersey_Number         16642 non-null  int64 
 10  Contract_Valid_Until  16642 non-null  int64 
 11  Height                16642 non-null  object
 12  Weight_lb             16642 non-null  int64 
 13  Release_Clause        16642 non-null  int64 
 14  Value                 16642 non-null  int64 
 15  Wage                  16642 non-null

In [9]:
def intt(x):
    li = x.split("'")
    lit = float(li[0])*30.48 + float(li[1])*2.54
    return lit


fifa["hight_cm"]= fifa["Height"].apply(lambda x:round(intt(x),2))
fifa.head()

ID               Name  Age Nationality  Overall                 Club  \
0  158023           L. Messi   31   Argentina       94         FC Barcelona   
1   20801  Cristiano Ronaldo   33    Portugal       94             Juventus   
2  190871          Neymar Jr   26      Brazil       92  Paris Saint-Germain   
3  193080             De Gea   27       Spain       91    Manchester United   
4  192985       K. De Bruyne   27     Belgium       91      Manchester City   

  Preferred_Foot       Work_Rate Position  Jersey_Number  \
0           Left  Medium/ Medium       RF             10   
1          Right       High/ Low       ST              7   
2          Right    High/ Medium       LW             10   
3          Right  Medium/ Medium       GK              1   
4          Right      High/ High      RCM              7   

   Contract_Valid_Until Height  Weight_lb  Release_Clause   Value  Wage  \
0                  2021    5'7        159          226500  110500   565   
1                  2022    6'2        183          127100   77000   405   
2                  2022    5'9        150          228100  118500   290   
3                  2020    6'4        168          138600   72000   260   
4                  2023   5'11        154          196400  102000   355   

   hight_cm  
0    170.18  
1    187.96  
2    175.26  
3    193.04  
4    180.34

In [ ]:
# Pandas를 사용할 때, 숫자가 지수표현식(과학적표기법)으로 나와서 불편할 때가 있습니다.
# 이럴 때는 Pandas display option을 바꿔주면 실수 표현이 가능합니다.
pd.options.display.float_format = '{:.5f}'.format
# 다시 원래대로 옵션을 변경하고 싶을 때는 아래 명령어를 사용하면 됩니다.
pd.reset_option('display.float_format')


# concat, merge, join

### 데이터프레임 결합
* 판다스 join() 메소드는 merge() 함수를 기반으로 만들어졌기 때문에 기본 작동 방식이 서로 비슷
* join() 메소드는 두 데이터프레임의 행 인덱스를 기준으로 결합하는 점에서 merge와 차이가 있으나 on=keys 옵션 설정하면 열을 기준으로 결합


In [269]:
data1 = np.arange(1,21).reshape(4,5)
data2 = np.arange(11,31).reshape(5,4)
df1 = pd.DataFrame(data1, columns=list("abcde"))
df2 = pd.DataFrame(data2,columns=list("dehi"))
print(df1,'\n')
print(df2)

    a   b   c   d   e
0   1   2   3   4   5
1   6   7   8   9  10
2  11  12  13  14  15
3  16  17  18  19  20 

    d   e   h   i
0  11  12  13  14
1  15  16  17  18
2  19  20  21  22
3  23  24  25  26
4  27  28  29  30


In [270]:
#join
df1.join(df2, lsuffix='_a',rsuffix='_b')

a   b   c  d_a  e_a  d_b  e_b   h   i
0   1   2   3    4    5   11   12  13  14
1   6   7   8    9   10   15   16  17  18
2  11  12  13   14   15   19   20  21  22
3  16  17  18   19   20   23   24  25  26

In [271]:
df1.join(df2, lsuffix='_a',rsuffix='_b',how='inner')

a   b   c  d_a  e_a  d_b  e_b   h   i
0   1   2   3    4    5   11   12  13  14
1   6   7   8    9   10   15   16  17  18
2  11  12  13   14   15   19   20  21  22
3  16  17  18   19   20   23   24  25  26

In [272]:
df1.join(df2, lsuffix='_a',rsuffix='_b',how='outer')

a     b     c   d_a   e_a  d_b  e_b   h   i
0   1.0   2.0   3.0   4.0   5.0   11   12  13  14
1   6.0   7.0   8.0   9.0  10.0   15   16  17  18
2  11.0  12.0  13.0  14.0  15.0   19   20  21  22
3  16.0  17.0  18.0  19.0  20.0   23   24  25  26
4   NaN   NaN   NaN   NaN   NaN   27   28  29  30

In [273]:
df1['ind']=[1,2,3,4]
df2['ind'] = [1,2,3,4,5]
df1

a   b   c   d   e  ind
0   1   2   3   4   5    1
1   6   7   8   9  10    2
2  11  12  13  14  15    3
3  16  17  18  19  20    4

In [274]:
df1.set_index('ind').join(df2.set_index('ind'),lsuffix='_a',rsuffix='_b')

a   b   c  d_a  e_a  d_b  e_b   h   i
ind                                        
1     1   2   3    4    5   11   12  13  14
2     6   7   8    9   10   15   16  17  18
3    11  12  13   14   15   19   20  21  22
4    16  17  18   19   20   23   24  25  26

In [486]:
df_s1 = pd.read_excel('dataset/stock price.xlsx',engine='openpyxl',index_col ='id')
df_s2 = pd.read_excel('dataset/stock valuation.xlsx',engine='openpyxl',index_col ='id')
print(df_s1,'\n')
print(df_s2)


       stock_name          value   price
id                                      
128940       한미약품   59385.666667  421000
130960     CJ E&M   58540.666667   98900
138250      엔에스쇼핑   14558.666667   13200
139480        이마트  239230.833333  254500
142280     녹십자엠에스     468.833333   10200
145990        삼양사   82750.000000   82000
185750        종근당   40293.666667  100500
192400      쿠쿠홀딩스  179204.666667  177500
199800         툴젠   -2514.333333  115400
204210     모두투어리츠    3093.333333    3475 

             name           eps     bps        per       pbr
id                                                          
130960     CJ E&M   6301.333333   54068  15.695091  1.829178
136480         하림    274.166667    3551  11.489362  0.887074
138040    메리츠금융지주   2122.333333   14894   6.313806  0.899691
139480        이마트  18268.166667  295780  13.931338  0.860437
145990        삼양사   5741.000000  108090  14.283226  0.758627
161390      한국타이어   5648.500000   51341   7.453306  0.820007
181710  NHN엔터테인먼트 

In [487]:
df_j1 = df_s1.join(df_s2, how='left')
df_j1

stock_name          value   price    name           eps       bps  \
id                                                                         
128940       한미약품   59385.666667  421000     NaN           NaN       NaN   
130960     CJ E&M   58540.666667   98900  CJ E&M   6301.333333   54068.0   
138250      엔에스쇼핑   14558.666667   13200     NaN           NaN       NaN   
139480        이마트  239230.833333  254500     이마트  18268.166667  295780.0   
142280     녹십자엠에스     468.833333   10200     NaN           NaN       NaN   
145990        삼양사   82750.000000   82000     삼양사   5741.000000  108090.0   
185750        종근당   40293.666667  100500     종근당   3990.333333   40684.0   
192400      쿠쿠홀딩스  179204.666667  177500     NaN           NaN       NaN   
199800         툴젠   -2514.333333  115400     NaN           NaN       NaN   
204210     모두투어리츠    3093.333333    3475  모두투어리츠     85.166667    5335.0   

              per       pbr  
id                           
128940        NaN       NaN  
130960  15.695091  1.829178  
138250        NaN       NaN  
139480  13.931338  0.860437  
142280        NaN       NaN  
145990  14.283226  0.758627  
185750  25.185866  2.470259  
192400        NaN       NaN  
199800        NaN       NaN  
204210  40.802348  0.651359

In [488]:
df_j1 = df_s1.join(df_s2, how='right')
df_j1

stock_name          value     price       name           eps     bps  \
id                                                                            
130960     CJ E&M   58540.666667   98900.0     CJ E&M   6301.333333   54068   
136480        NaN            NaN       NaN         하림    274.166667    3551   
138040        NaN            NaN       NaN    메리츠금융지주   2122.333333   14894   
139480        이마트  239230.833333  254500.0        이마트  18268.166667  295780   
145990        삼양사   82750.000000   82000.0        삼양사   5741.000000  108090   
161390        NaN            NaN       NaN      한국타이어   5648.500000   51341   
181710        NaN            NaN       NaN  NHN엔터테인먼트   2110.166667   78434   
185750        종근당   40293.666667  100500.0        종근당   3990.333333   40684   
204210     모두투어리츠    3093.333333    3475.0     모두투어리츠     85.166667    5335   
207940        NaN            NaN       NaN   삼성바이오로직스   4644.166667   60099   

              per       pbr  
id                           
130960  15.695091  1.829178  
136480  11.489362  0.887074  
138040   6.313806  0.899691  
139480  13.931338  0.860437  
145990  14.283226  0.758627  
161390   7.453306  0.820007  
181710  30.755864  0.827447  
185750  25.185866  2.470259  
204210  40.802348  0.651359  
207940  89.790059  6.938551

In [493]:
df_j2 = df_s1.join(df_s2, how='inner')#교집합, 서로 공통돤 것만 나옴
df_j2

stock_name          value   price    name           eps     bps  \
id                                                                       
130960     CJ E&M   58540.666667   98900  CJ E&M   6301.333333   54068   
139480        이마트  239230.833333  254500     이마트  18268.166667  295780   
145990        삼양사   82750.000000   82000     삼양사   5741.000000  108090   
185750        종근당   40293.666667  100500     종근당   3990.333333   40684   
204210     모두투어리츠    3093.333333    3475  모두투어리츠     85.166667    5335   

              per       pbr  
id                           
130960  15.695091  1.829178  
139480  13.931338  0.860437  
145990  14.283226  0.758627  
185750  25.185866  2.470259  
204210  40.802348  0.651359

## 과제(7/4)
Q. df_j2에서 중복된 칼럼을 삭제한 후 블린 인덱싱을 이용하여 eps가 3000 보다 적거나 stock_name이 이마트인 데이터를 선택하여 
데이터프레임을 생성하고 df_j3 이름으로 저장 및 출력하세요.(단, '<' 와 '==' 를 반드시 사용해야 함)


In [442]:
dfj3= df_j2.drop(["name"], axis = 1)
dfj3
df_j3 = dfj3[(dfj3["eps"] < 3000) | (dfj3["stock_name"]=="이마트")]
df_j3

stock_name          value   price           eps     bps        per  \
id                                                                          
139480        이마트  239230.833333  254500  18268.166667  295780  13.931338   
204210     모두투어리츠    3093.333333    3475     85.166667    5335  40.802348   

             pbr  
id                
139480  0.860437  
204210  0.651359

### 데이터프레임 병합
* merge() 함수는 sql의 join 명령과 비슷한 방식으로 어떤 기준에 의해 두 데이터프레임을 병합하는 개념
* 기준이 되는 열이나 인덱스를 키라고 부름
* 키가 되는 열이나 인덱스는 반드시 양쪽 데이터프레임에 모두 존재함


In [445]:
data1 = np.arange(1,21).reshape(4,5)
data2 = np.arange(11,31).reshape(5,4)
df1 = pd.DataFrame(data1, columns=list("abcde"))
df2 = pd.DataFrame(data2,columns=list("dehi"))
print(df1,'\n')
print(df2)

    a   b   c   d   e
0   1   2   3   4   5
1   6   7   8   9  10
2  11  12  13  14  15
3  16  17  18  19  20 

    d   e   h   i
0  11  12  13  14
1  15  16  17  18
2  19  20  21  22
3  23  24  25  26
4  27  28  29  30


In [446]:
df1['ind']=[1,2,3,4]
df2['ind'] = [1,2,3,4,5]
df1

a   b   c   d   e  ind
0   1   2   3   4   5    1
1   6   7   8   9  10    2
2  11  12  13  14  15    3
3  16  17  18  19  20    4

In [447]:
pd.merge(df1,df2,on='ind',how='outer')

a     b     c   d_x   e_x  ind  d_y  e_y   h   i
0   1.0   2.0   3.0   4.0   5.0    1   11   12  13  14
1   6.0   7.0   8.0   9.0  10.0    2   15   16  17  18
2  11.0  12.0  13.0  14.0  15.0    3   19   20  21  22
3  16.0  17.0  18.0  19.0  20.0    4   23   24  25  26
4   NaN   NaN   NaN   NaN   NaN    5   27   28  29  30

### 과제(7/4):
df1, df2에 대하여 다음을 수행하세요.
- on = 'None'으로 교집합을 만들어 출력하세요.
- 첫번째 df (stock_name), 두번째 df (name)을 기준으로 합집합(left)을 만들어 출력하세요.
- 오른쪽 df를 기준으로 합친 후, 출력하세요.
- df1에서 가격이 50000 미만인 데이터만 출력하세요.
- df2에서 per가 30보다 크고 pbr이 0.8보다 큰 데이터를 모두 출력하세요.

In [10]:
df1 = pd.read_csv('dataset/stock_price.csv',encoding = "cp949",index_col ='id')
df2 = pd.read_csv('dataset/stock_valuation.csv',encoding = "cp949",index_col ='id')
print(df1,'\n')
print(df2)

       stock_name          value   price
id                                      
128940       한미약품   59385.666670  421000
130960     CJ E&M   58540.666670   98900
138250      엔에스쇼핑   14558.666670   13200
139480        이마트  239230.833300  254500
142280     녹십자엠에스     468.833333   10200
145990        삼양사   82750.000000   82000
185750        종근당   40293.666670  100500
192400      쿠쿠홀딩스  179204.666700  177500
199800         툴젠   -2514.333333  115400
204210     모두투어리츠    3093.333333    3475 

             name           eps     bps        per       pbr
id                                                          
130960     CJ E&M   6301.333333   54068  15.695091  1.829178
136480         하림    274.166667    3551  11.489362  0.887074
138040    메리츠금융지주   2122.333333   14894   6.313806  0.899691
139480        이마트  18268.166670  295780  13.931338  0.860437
145990        삼양사   5741.000000  108090  14.283226  0.758627
161390      한국타이어   5648.500000   51341   7.453306  0.820007
181710  NHN엔터테인먼트 

In [11]:
#1)
df3 = df1.join(df2, on = None, how ='inner')
df3

stock_name          value   price    name           eps     bps  \
id                                                                       
130960     CJ E&M   58540.666670   98900  CJ E&M   6301.333333   54068   
139480        이마트  239230.833300  254500     이마트  18268.166670  295780   
145990        삼양사   82750.000000   82000     삼양사   5741.000000  108090   
185750        종근당   40293.666670  100500     종근당   3990.333333   40684   
204210     모두투어리츠    3093.333333    3475  모두투어리츠     85.166667    5335   

              per       pbr  
id                           
130960  15.695091  1.829178  
139480  13.931338  0.860437  
145990  14.283226  0.758627  
185750  25.185866  2.470259  
204210  40.802348  0.651359

In [27]:
#2)첫번째 df (stock_name), 두번째 df (name)을 기준으로 합집합(left)을 만들어 출력하세요.
df4 = pd.merge(df1, df2, left_on = "stock_name", right_on="name", how = "left")
df4

stock_name          value   price    name           eps       bps  \
0       한미약품   59385.666670  421000     NaN           NaN       NaN   
1     CJ E&M   58540.666670   98900  CJ E&M   6301.333333   54068.0   
2      엔에스쇼핑   14558.666670   13200     NaN           NaN       NaN   
3        이마트  239230.833300  254500     이마트  18268.166670  295780.0   
4     녹십자엠에스     468.833333   10200     NaN           NaN       NaN   
5        삼양사   82750.000000   82000     삼양사   5741.000000  108090.0   
6        종근당   40293.666670  100500     종근당   3990.333333   40684.0   
7      쿠쿠홀딩스  179204.666700  177500     NaN           NaN       NaN   
8         툴젠   -2514.333333  115400     NaN           NaN       NaN   
9     모두투어리츠    3093.333333    3475  모두투어리츠     85.166667    5335.0   

         per       pbr  
0        NaN       NaN  
1  15.695091  1.829178  
2        NaN       NaN  
3  13.931338  0.860437  
4        NaN       NaN  
5  14.283226  0.758627  
6  25.185866  2.470259  
7        NaN       NaN  
8        NaN       NaN  
9  40.802348  0.651359

In [25]:
df6 = df1.join( df2, how = "left")
df6

stock_name          value   price    name           eps       bps  \
id                                                                         
128940       한미약품   59385.666670  421000     NaN           NaN       NaN   
130960     CJ E&M   58540.666670   98900  CJ E&M   6301.333333   54068.0   
138250      엔에스쇼핑   14558.666670   13200     NaN           NaN       NaN   
139480        이마트  239230.833300  254500     이마트  18268.166670  295780.0   
142280     녹십자엠에스     468.833333   10200     NaN           NaN       NaN   
145990        삼양사   82750.000000   82000     삼양사   5741.000000  108090.0   
185750        종근당   40293.666670  100500     종근당   3990.333333   40684.0   
192400      쿠쿠홀딩스  179204.666700  177500     NaN           NaN       NaN   
199800         툴젠   -2514.333333  115400     NaN           NaN       NaN   
204210     모두투어리츠    3093.333333    3475  모두투어리츠     85.166667    5335.0   

              per       pbr  
id                           
128940        NaN       NaN  
130960  15.695091  1.829178  
138250        NaN       NaN  
139480  13.931338  0.860437  
142280        NaN       NaN  
145990  14.283226  0.758627  
185750  25.185866  2.470259  
192400        NaN       NaN  
199800        NaN       NaN  
204210  40.802348  0.651359

In [544]:
#3)
df5 = df1.join(df2, lsuffix = "stock_name", rsuffix="name", how = "right")
df5

stock_name          value     price       name           eps     bps  \
id                                                                            
130960     CJ E&M   58540.666670   98900.0     CJ E&M   6301.333333   54068   
136480        NaN            NaN       NaN         하림    274.166667    3551   
138040        NaN            NaN       NaN    메리츠금융지주   2122.333333   14894   
139480        이마트  239230.833300  254500.0        이마트  18268.166670  295780   
145990        삼양사   82750.000000   82000.0        삼양사   5741.000000  108090   
161390        NaN            NaN       NaN      한국타이어   5648.500000   51341   
181710        NaN            NaN       NaN  NHN엔터테인먼트   2110.166667   78434   
185750        종근당   40293.666670  100500.0        종근당   3990.333333   40684   
204210     모두투어리츠    3093.333333    3475.0     모두투어리츠     85.166667    5335   
207940        NaN            NaN       NaN   삼성바이오로직스   4644.166667   60099   

              per       pbr  
id                           
130960  15.695091  1.829178  
136480  11.489362  0.887074  
138040   6.313806  0.899691  
139480  13.931338  0.860437  
145990  14.283226  0.758627  
161390   7.453306  0.820007  
181710  30.755864  0.827447  
185750  25.185866  2.470259  
204210  40.802348  0.651359  
207940  89.790059  6.938551

In [549]:
#4)
df1[df1["price"]<50000]

stock_name         value  price
id                                    
138250      엔에스쇼핑  14558.666670  13200
142280     녹십자엠에스    468.833333  10200
204210     모두투어리츠   3093.333333   3475

In [557]:
#5)df2에서 per가 30보다 크고 pbr이 0.8보다 큰 데이터를 모두 출력하세요.
df2[(df2["per"]>30)&(df2["pbr"]>0.8)]

name          eps    bps        per       pbr
id                                                        
181710  NHN엔터테인먼트  2110.166667  78434  30.755864  0.827447
207940   삼성바이오로직스  4644.166667  60099  89.790059  6.938551

In [333]:

col1 = np.random.uniform(10,20,size=10) #10~20 사이 실수 10개
col2 = np.random.uniform(10,20,size = 10)
sr1=pd.Series(col1,name='col1')
sr2=pd.Series(col2,name='col2')
df = pd.concat([sr1,sr2],axis=1)
df = df.round({'col1':2,'col2':1})
df.head()

col1  col2
0  10.81  16.3
1  10.85  17.1
2  10.36  18.5
3  10.69  11.4
4  15.75  14.0

In [339]:
#col2를 정수로 변환하여 출력
df1 = df.copy()
df1 = df.col2.astype(int)
df1

0    16
1    17
2    18
3    11
4    14
5    14
6    11
7    18
8    10
9    19
Name: col2, dtype: int32

In [340]:
df2 = df.copy()
df2.iloc[1:3,:]=np.nan
df2.head()

col1  col2
0  10.81  16.0
1    NaN   NaN
2    NaN   NaN
3  10.69  11.0
4  15.75  14.0

In [344]:
#null값을 0으로 채우기
df2.replace(np.nan,0,inplace=True)
df2 = df2.astype({'col1':'int','col2':'int'})
df2.head()

col1  col2
0    10    16
1     0     0
2     0     0
3    10    11
4    15    14

In [347]:
df2['col0'] = list('abcdefghij')
df2.set_index('col0',inplace = True)
df2.head()

col1  col2
col0            
a       10    16
b        0     0
c        0     0
d       10    11
e       15    14

In [349]:
df2.isna().sum()

col1    0
col2    0
dtype: int64

In [351]:
df2 = df2.reset_index()
df2.head()

index col0  col1  col2
0      0    a    10    16
1      1    b     0     0
2      2    c     0     0
3      3    d    10    11
4      4    e    15    14

In [352]:
#reindex
df3 = df2.reindex(index=[1,2],columns=['col1','col2'])
df3

col1  col2
1     0     0
2     0     0

In [355]:
df4 = df2.reindex(index = [3,4],columns=['col1','col2'])
df4

col1  col2
3    10    11
4    15    14

# scaling(정규화, 표준화)

In [360]:
from sklearn.datasets import load_boston
import warnings
warnings.filterwarnings('ignore')
boston = load_boston()
print(boston.keys())
print(boston.feature_names)
df_boston = pd.DataFrame(boston.data,columns=boston.feature_names)
df_boston['target']=boston.target
df_boston.head()

dict_keys(['data', 'target', 'feature_names', 'DESCR', 'filename', 'data_module'])
['CRIM' 'ZN' 'INDUS' 'CHAS' 'NOX' 'RM' 'AGE' 'DIS' 'RAD' 'TAX' 'PTRATIO'
 'B' 'LSTAT']


CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   

   PTRATIO       B  LSTAT  target  
0     15.3  396.90   4.98    24.0  
1     17.8  396.90   9.14    21.6  
2     17.8  392.83   4.03    34.7  
3     18.7  394.63   2.94    33.4  
4     18.7  396.90   5.33    36.2

In [366]:
from sklearn.preprocessing import MinMaxScaler
#0~1사이
scaler = MinMaxScaler()
df_boston['RM']=scaler.fit_transform(df_boston[['RM']])
df_boston[['RM']].describe()

RM
count  506.000000
mean     0.521869
std      0.134627
min      0.000000
25%      0.445392
50%      0.507281
75%      0.586798
max      1.000000

In [373]:
from sklearn.preprocessing import StandardScaler#표준분포
scaler = StandardScaler()
df_boston['INDUS'] = scaler.fit_transform(df_boston[['INDUS']])
df_boston[['INDUS']].describe()

INDUS
count  5.060000e+02
mean   1.040560e-16
std    1.000990e+00
min   -1.557842e+00
25%   -8.676906e-01
50%   -2.110985e-01
75%    1.015999e+00
max    2.422565e+00

# encoding: LableEncoder, OneHotEncoder

In [65]:
import pandas as pd
import numpy as np
import seaborn as sns

titanic = sns.load_dataset('titanic')
#특정 컬럼의 값을 변경
titanic['sex'] = titanic.sex.replace({'male' : 0, 'female' : 1})
#열이름 변경
titanic.rename(columns = {'sex':'gender', 'age':'age_cat','class':'c_level'},inplace = True)
#컬럼명을 대문자로 변경
titanic.columns = map(lambda x:str(x).upper(), titanic.columns)
#컬럼명을 소문자로 변경
titanic.rename(columns = str.lower, inplace = True)
df= titanic[['survived','pclass','gender','c_level']]
df.head()

survived  pclass  gender c_level
0         0       3       0   Third
1         1       1       1   First
2         1       3       1   Third
3         1       1       1   First
4         0       3       0   Third

In [79]:
import warnings
warnings.filterwarnings('ignore')
df['gender'].replace([1,0],['female','male'],inplace = True)
df.head()

survived  pclass  gender c_level
0         0       3    male   Third
1         1       1  female   First
2         1       3  female   Third
3         1       1  female   First
4         0       3    male   Third

In [67]:
df1 = df.copy()

In [68]:
#인코딩 - labelEncoder, 숫자로 나오는 요소 값의 크기가 데이터 분석 결과에 영향을 줄 수 있다. 
#요소를 표현하는 숫자의 크기는 아무 상관이 없다.
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
features =['gender','c_level']
for f in features:
    df1[f] = le.fit_transform(df1[f])
df1.head()

survived  pclass  gender  c_level
0         0       3       1        2
1         1       1       0        0
2         1       3       0        2
3         1       1       0        0
4         0       3       1        2

In [69]:
df1.c_level.value_counts()

2    491
0    216
1    184
Name: c_level, dtype: int64

In [80]:
# 인코딩 - OneHotEncoder(aka. 희소 입력, 희소행렬)
df2 = df.copy()
df3 = df2[['gender','c_level']]
pd.get_dummies(df3).head()

gender_female  gender_male  c_level_First  c_level_Second  c_level_Third
0              0            1              0               0              1
1              1            0              1               0              0
2              1            0              0               0              1
3              1            0              1               0              0
4              0            1              0               0              1

### 과제(7/5)

- titanic 데이터셋에서 embarked, who, adult_male에 대하여 LabelEncoding 하세요.
- titanic 데이터셋에서 embarked, who, adult_male에 대하여 OneHotEncoding 하세요.

In [10]:
titanic = sns.load_dataset('titanic')
df = titanic[['embarked','who','adult_male']]
df.head()

embarked    who  adult_male
0        S    man        True
1        C  woman       False
2        S  woman       False
3        S  woman       False
4        S    man        True

In [11]:
df1 = df.copy()
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder() # < - 객체
features =['embarked','who','adult_male']
for f in features:
    df1[f] = le.fit_transform(df1[f])
df1.head()

embarked  who  adult_male
0         2    1           1
1         0    2           0
2         2    2           0
3         2    2           0
4         2    1           1

In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 3 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   embarked    889 non-null    object
 1   who         891 non-null    object
 2   adult_male  891 non-null    bool  
dtypes: bool(1), object(2)
memory usage: 14.9+ KB


In [17]:
df.adult_male = df.adult_male.astype(str)

C:\Users\admin\AppData\Local\Temp\ipykernel_4180\227025549.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.adult_male = df.adult_male.astype(str)


In [18]:
#OnHotEncoding

df2 = df.copy()
df3 = df2[['embarked','who','adult_male']]
pd.get_dummies(df3).head()

embarked_C  embarked_Q  embarked_S  who_child  who_man  who_woman  \
0           0           0           1          0        1          0   
1           1           0           0          0        0          1   
2           0           0           1          0        0          1   
3           0           0           1          0        0          1   
4           0           0           1          0        1          0   

   adult_male_False  adult_male_True  
0                 0                1  
1                 1                0  
2                 1                0  
3                 1                0  
4                 0                1

### Q1. titanic 데이터셋에서 age, fare 2개 열을 선택하여 데이터프레임 df를 만드세요

In [3]:
import pandas as pd
import seaborn as sns

tin = sns.load_dataset('titanic')
tin

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

In [4]:
df = tin.loc[:,['age','fare']]
print(df.head())

    age     fare
0  22.0   7.2500
1  38.0  71.2833
2  26.0   7.9250
3  35.0  53.1000
4  35.0   8.0500


### Q2.  10을 더하는 함수 add_10 과 두 객체의 합 add_two_obj 함수를 정의한 후 결과값이 각각 20과 20을 출력하세요.


In [5]:
def add_10(a):
    return a + 10
def add_two_obj(a,b):
    return a+b
print(add_10(10))
print(add_two_obj(10, 10))

20
20


### Q3. 10을 곱하는 사용자 함수와 두 객체의 나눗셈을 수행하는 사용자 함수를 정의한 후 결과값이 각각 200과 10을 출력하세요.


In [6]:
def mul_10(a):
    return a*10
def div_two_obj(a,b):
    return a/b
print(mul_10(20))
print(int(div_two_obj(20,2)))

200
10


### Q4. df.age 에 add_10 함수를 적용해서 출력하세요.


In [7]:

sr= df.age.apply(add_10)
sr.head()

0    32.0
1    48.0
2    36.0
3    45.0
4    45.0
Name: age, dtype: float64

### Q5. 시리즈 객체 + 숫자를 연산하는 방식으로 df.age에 10을 더한 후 sr2에 저장하고 출력하세요.


In [9]:
sr2 = df.age + 10
sr2.head()

0    32.0
1    48.0
2    36.0
3    45.0
4    45.0
Name: age, dtype: float64

### Q6. 람다 함수를 활용해서 위와 같은 연산을 수행하고 결과를 sr3로 출력하세요.


In [10]:
x = lambda a : a + 10
sr3 = df.age.apply(x)
sr3.head()

0    32.0
1    48.0
2    36.0
3    45.0
4    45.0
Name: age, dtype: float64

### Q7. df의 'age', 'fare' 칼럼에 10을 덧셈연산하는 사용자 함수를 정의하고 applymap()으로 각 사용자 함수를 매핑한 후 출력하세요.


In [14]:
def add_10(a):
    return a + 10
df1 = df.copy()

df_map= df1.applymap(add_10)
df_map.head()

age     fare
0  32.0  17.2500
1  48.0  81.2833
2  36.0  17.9250
3  45.0  63.1000
4  45.0  18.0500

### Q8. df_map의 [0,0]을 NaN으로 변경한 후 missing_value 함수를 적용해서 불린 시리즈를 반환하세요.


In [100]:
df_map.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   age     714 non-null    float64
 1   fare    891 non-null    float64
dtypes: float64(2)
memory usage: 14.0 KB


In [169]:
df_map.iloc[0,0] = None

def misssing_value(x):
    return x.isnull()

df3 = df_map.apply(misssing_value)
print(df3.iloc[:,0].head(),"\n")
print(df3.iloc[:,1].head())

0     True
1    False
2    False
3    False
4    False
Name: age, dtype: bool 

0    False
1    False
2    False
3    False
4    False
Name: fare, dtype: bool


In [16]:
df_map.iloc[0,0] = None

def misssing_value(series):
    return series.isnull()

df3 = df_map.apply(misssing_value, axis = 0)
print(df3)


       age   fare
0     True  False
1    False  False
2    False  False
3    False  False
4    False  False
..     ...    ...
886  False  False
887  False  False
888   True  False
889  False  False
890  False  False

[891 rows x 2 columns]


### Q9. titanic 데이터셋을 seaborn에서 로딩한 후 블린 인덱싱을 이용하여 아래 사항들을 수행하세요.
- 나이가 10대(10~19세)인 승객만 따로 선택
- 나이가 10세 미만(0~9세)이고 여성인 승객만 따로 선택
- 나이가 10세 미만(0~9세) 또는 60세 이상인 승객의 age, sex, alone 열만 선택


In [17]:
import pandas as pd
import seaborn as sns

tin = sns.load_dataset('titanic')
tin.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

In [24]:
#mask1 = (tin["age"]<20)&(tin["age"]>=10)
#tin1 = tin.loc[mask1,:]
#tin1

#tin[(tin["age"]<20)&(tin["age"]>=10)]

#tin[(tin["age"]<10)&(tin["sex"] == "female")]

df3 = tin[["age","sex","alone"]]
df3[(df3["age"]<10)|(df3["age"]>=60)].head()



age     sex  alone
7    2.0    male  False
10   4.0  female  False
16   2.0    male  False
24   8.0  female  False
33  66.0    male   True

### Q10. 나이가 10세 미만(0~9세) 또는 60세 이상인 승객의 alone 열만 선택한 후 값을 True로 변경하세요


In [32]:
df1=tin[(tin["age"]<10)|(tin["age"]>=60)]
df2 = df1.loc[:,["alone"]]
print(df2.head())
df2["alone"] = True
df2.head()

    alone
7   False
10  False
16  False
24  False
33   True


alone
7    True
10   True
16   True
24   True
33   True

### Q11. 최대값 - 최소값을 반환하는 사용자 함수를 min_max로 정의하고 age, fare 열로 구성되는 타이타닉 데이터프레임df 에 적용하여 결과와 자료형을 출력하세요. source : titanic = sns.load_dataset('titanic')


In [141]:
import pandas as pd
import seaborn as sns
import numpy as np

titanic = sns.load_dataset('titanic')

In [145]:
titanic.isnull().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

In [299]:
df = titanic[["age","fare"]]
df.head()

age     fare
0  22.0   7.2500
1  38.0  71.2833
2  26.0   7.9250
3  35.0  53.1000
4  35.0   8.0500

In [300]:
def min_max(x):
    print(min(x))
    print(max(x))
df.apply(min_max)

0.42
80.0
0.0
512.3292


age     None
fare    None
dtype: object

### Q12. df 에서 'age' 칼럼과 'fare'칼럼을 더한 가격을 리턴하는 사용자 함수를 정의한 후 df['add'] 변수를 생성하여 출력하세요.(lambda 함수 사용)


In [301]:
def adds(a,b):
    return a+b
df["add"] = df.apply(lambda x:adds(x["age"], x["fare"]), axis = 1)
df.head()

C:\Users\admin\AppData\Local\Temp\ipykernel_1672\221001242.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["add"] = df.apply(lambda x:adds(x["age"], x["fare"]), axis = 1)


age     fare       add
0  22.0   7.2500   29.2500
1  38.0  71.2833  109.2833
2  26.0   7.9250   33.9250
3  35.0  53.1000   88.1000
4  35.0   8.0500   43.0500

In [107]:
df5= df.iloc[:,0] + df.iloc[:,1]
df5

0       29.2500
1      109.2833
2       33.9250
3       88.1000
4       43.0500
         ...   
886     40.0000
887     49.0000
888         NaN
889     56.0000
890     39.7500
Length: 891, dtype: float64

### Q13. 'age', 'fare' 두개 칼럼으로 구성되는 titanic 데이터프레임에서 [0,1]의 값을 NaN을 변경한 후 df 이름으로 출력하세요.


In [168]:
import pandas as pd
import seaborn as sns
import numpy as np

titanic = sns.load_dataset('titanic')

In [169]:
titanic.isnull().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

In [316]:
df = titanic[["age","fare"]]
df.head()

age     fare
0  22.0   7.2500
1  38.0  71.2833
2  26.0   7.9250
3  35.0  53.1000
4  35.0   8.0500

In [328]:
df[df["fare"]==df.iloc[0,1]].count()

age     10
fare    13
dtype: int64

In [308]:
df1 = df.copy()
df1.iloc[0,1] = np.nan
df1.head()

age     fare       add
0  22.0      NaN   29.2500
1  38.0  71.2833  109.2833
2  26.0   7.9250   33.9250
3  35.0  53.1000   88.1000
4  35.0   8.0500   43.0500

In [309]:
df1.isnull().sum()

age     177
fare      1
add     177
dtype: int64

### Q14. 각 열의 NaN 찾기, 각 열의 NaN 개수 반환, 데이터프레임의 총 NaN 개수를 반환하는 사용자 함수를 정의한 후 df.pipe 메소드로 함수 매핑한 결과를 모두 출력하세요.


In [286]:
def nan(x):
    print(x["age"].isnull())
    print(x["age"].isnull().sum())
    print(x["fare"].isnull())
    print(x["fare"].isnull().sum())
    print(x["age"].isnull().sum()+x["fare"].isnull().sum())
    return x
df1.pipe(nan)

0      False
1      False
2      False
3      False
4      False
       ...  
886    False
887    False
888     True
889    False
890    False
Name: age, Length: 891, dtype: bool
177
0       True
1      False
2      False
3      False
4      False
       ...  
886    False
887    False
888    False
889    False
890    False
Name: fare, Length: 891, dtype: bool
1
178


age     fare
0    22.0      NaN
1    38.0  71.2833
2    26.0   7.9250
3    35.0  53.1000
4    35.0   8.0500
..    ...      ...
886  27.0  13.0000
887  19.0  30.0000
888   NaN  23.4500
889  26.0  30.0000
890  32.0   7.7500

[891 rows x 2 columns]

In [329]:
def nan(x):
    print(x.isnull().sum())
    return x
df.pipe(nan)

age     177
fare      0
dtype: int64


age     fare
0    22.0   7.2500
1    38.0  71.2833
2    26.0   7.9250
3    35.0  53.1000
4    35.0   8.0500
..    ...      ...
886  27.0  13.0000
887  19.0  30.0000
888   NaN  23.4500
889  26.0  30.0000
890  32.0   7.7500

[891 rows x 2 columns]

### Q15. 타이타닉 df의 열 순서를 아래와 같이 재구성하고 출력하세요
* 열 선택 : ['pclass', 'sex', 'age', 'survived'].
* 열 이름의 리스트 만들기
* 열 이름을 알파벳 순으로 정렬하기
* 열 이름을 기존 순서의 정반대 역순으로 정렬하기
* 열 이름을 사용자가 정의한 임의의 순서로 재배치하기


In [48]:
df3 = titanic.copy()

In [54]:
df2 = df3[["pclass", "sex", "age", "survived"]]
df2

pclass     sex   age  survived
0         3    male  22.0         0
1         1  female  38.0         1
2         3  female  26.0         1
3         1  female  35.0         1
4         3    male  35.0         0
..      ...     ...   ...       ...
886       2    male  27.0         0
887       1  female  19.0         1
888       3  female   NaN         0
889       1    male  26.0         1
890       3    male  32.0         0

[891 rows x 4 columns]

In [61]:
liti = list(df2)
liti

['pclass', 'sex', 'age', 'survived']

In [63]:
liti.sort()
liti

['age', 'pclass', 'sex', 'survived']

In [66]:
liti.reverse()
liti

['survived', 'sex', 'pclass', 'age']

In [72]:
df4 = df2[["sex","age","pclass","survived"]]
liti = list(df4)
liti

['sex', 'age', 'pclass', 'survived']

In [150]:
columns = ["sex","age","pclass","survived"]
df_customed = df3[columns]
df_customed.head()

sex   age  pclass  survived
0    male  22.0       3         0
1  female  38.0       1         1
2  female  26.0       3         1
3  female  35.0       1         1
4    male  35.0       3         0

### Q16. 함수 매핑을 이용하여 age에 각각 10을 곱하고 5를 더한 후 다시 더한 결과 값이 출력되게 하세요.


In [29]:
# (a * b) + (a + c)
# 데이터 가져오기
import seaborn as sns

# titanic 데이터셋에서 

titanic = sns.load_dataset('titanic')
# print(titanic.head())
df = titanic.loc[:, ['age','fare']]
print(df.head())


    age     fare
0  22.0   7.2500
1  38.0  71.2833
2  26.0   7.9250
3  35.0  53.1000
4  35.0   8.0500


In [37]:
def mul_add(a,b,c):
    return (a*b)+(a+c)

df["age"].apply(mul_add,b = 10, c=5)

0      247.0
1      423.0
2      291.0
3      390.0
4      390.0
       ...  
886    302.0
887    214.0
888      NaN
889    291.0
890    357.0
Name: age, Length: 891, dtype: float64

In [32]:
def mulmu(x):
    return (x*10) +(x+5)
df.age.apply(mulmu)


0      247.0
1      423.0
2      291.0
3      390.0
4      390.0
       ...  
886    302.0
887    214.0
888      NaN
889    291.0
890    357.0
Name: age, Length: 891, dtype: float64

### Q17. titanic.fare 데이터를 소수 2째자리 형태로 모두 출력하세요.
- 7.2500 -> 7.25

In [563]:
def soso(x):
    return round(x,2)
titanic.fare.apply(soso)

0       7.25
1      71.28
2       7.92
3      53.10
4       8.05
       ...  
886    13.00
887    30.00
888    23.45
889    30.00
890     7.75
Name: fare, Length: 891, dtype: float64

### Q18. df에서 fare와 age를 선택하여 100을 곱한 후 출력하세요


In [566]:
import pandas as pd

titanic = sns.load_dataset('titanic')
df = titanic
df.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

In [571]:
df[["fare","age"]].apply(lambda x:x*100).head()

fare     age
0   725.00  2200.0
1  7128.33  3800.0
2   792.50  2600.0
3  5310.00  3500.0
4   805.00  3500.0

### Q19. join() 메소드는 두 데이터프레임의 행 인덱스를 기준으로 결합한다. 2개의 주식데이터를 가져와서 join() 메소드로 아래와 같이 결합한 후 다음 사항을 수행하세요.
- df1과 df2의 교집합만 출력되도록 결합하여 df4에 저장하고 출력
- df4에서 중복된 칼럼을 삭제한 후 블린 인덱싱을 이용하여 eps가 3000 보다 적거나 stock_name이 이마트인 데이터를 선택하여 데이터프레임을 생성하고 df5 이름으로 저장 및 출력하세요.(단, '<' 와 '==' 를 반드시 사용해야 함)


In [573]:
import pandas as pd

# 주식 데이터를 가져와서 데이터프레임 만들기
df1 = pd.read_excel('./dataset/stock price.xlsx', index_col='id')
df2 = pd.read_excel('./dataset/stock valuation.xlsx', index_col='id')

# 데이터프레임 결합(join)
# df1의 행 인덱스를 기준으로 결합하는 how='left'옵션이 기본 적용
df3 = df1.join(df2)
print(df3)


       stock_name          value   price    name           eps       bps  \
id                                                                         
128940       한미약품   59385.666667  421000     NaN           NaN       NaN   
130960     CJ E&M   58540.666667   98900  CJ E&M   6301.333333   54068.0   
138250      엔에스쇼핑   14558.666667   13200     NaN           NaN       NaN   
139480        이마트  239230.833333  254500     이마트  18268.166667  295780.0   
142280     녹십자엠에스     468.833333   10200     NaN           NaN       NaN   
145990        삼양사   82750.000000   82000     삼양사   5741.000000  108090.0   
185750        종근당   40293.666667  100500     종근당   3990.333333   40684.0   
192400      쿠쿠홀딩스  179204.666667  177500     NaN           NaN       NaN   
199800         툴젠   -2514.333333  115400     NaN           NaN       NaN   
204210     모두투어리츠    3093.333333    3475  모두투어리츠     85.166667    5335.0   

              per       pbr  
id                           
128940        NaN       NaN

In [576]:
df4 = df1.join(df2,how = "inner")
df4

stock_name          value   price    name           eps     bps  \
id                                                                       
130960     CJ E&M   58540.666667   98900  CJ E&M   6301.333333   54068   
139480        이마트  239230.833333  254500     이마트  18268.166667  295780   
145990        삼양사   82750.000000   82000     삼양사   5741.000000  108090   
185750        종근당   40293.666667  100500     종근당   3990.333333   40684   
204210     모두투어리츠    3093.333333    3475  모두투어리츠     85.166667    5335   

              per       pbr  
id                           
130960  15.695091  1.829178  
139480  13.931338  0.860437  
145990  14.283226  0.758627  
185750  25.185866  2.470259  
204210  40.802348  0.651359

In [580]:
#df4에서 중복된 칼럼을 삭제한 후 블린 인덱싱을 이용하여 eps가 3000 보다 적거나 stock_name이 이마트인 데이터를 선택하여 데이터프레임을 생성하고 df5 이름으로 저장 및 출력하세요.(단, '<' 와 '==' 를 반드시 사용해야 함)
df5 = df4.drop("name",axis = 1)
df5
df6 = df5[(df5["eps"]<3000)|(df5["stock_name"] == "이마트")]
df6

stock_name          value   price           eps     bps        per  \
id                                                                          
139480        이마트  239230.833333  254500  18268.166667  295780  13.931338   
204210     모두투어리츠    3093.333333    3475     85.166667    5335  40.802348   

             pbr  
id                
139480  0.860437  
204210  0.651359

### Q20. 아래 df에서 중복 행을 제거한 후 df2에 저장하고 출력하세요.


In [581]:
import pandas as pd
df = pd.DataFrame({'c1':['a', 'a', 'b', 'a', 'b'],
                  'c2':[1, 1, 1, 2, 2],
                  'c3':[1, 1, 2, 2, 2]})
print(df)

  c1  c2  c3
0  a   1   1
1  a   1   1
2  b   1   2
3  a   2   2
4  b   2   2


In [583]:
df2 = df.drop_duplicates()
df2

c1  c2  c3
0  a   1   1
2  b   1   2
3  a   2   2
4  b   2   2

### 과제(7/5)

Q21~Q37을 수행하세요.

### Q21. Q20의 df에서 drop_duplicates() 메소드를 사용하여 c2, c3열을 기준으로 중복 행을 제거한 후 df3에 저장하고 출력하세요.


In [22]:
import pandas as pd
df = pd.DataFrame({'c1':['a', 'a', 'b', 'a', 'b'],
                  'c2':[1, 1, 1, 2, 2],
                  'c3':[1, 1, 2, 2, 2]})
print(df)

  c1  c2  c3
0  a   1   1
1  a   1   1
2  b   1   2
3  a   2   2
4  b   2   2


In [23]:
df.duplicated(['c2','c3'])

0    False
1     True
2    False
3    False
4     True
dtype: bool

In [102]:
df3 = df.drop_duplicates(["c2","c3"])
df3

c1  c2  c3
0  a   1   1
2  b   1   2
3  a   2   2

### Q33. './dataset/stock-data.csv'를 데이터프레임으로 불러와서 datetime64 자료형으로 변환한 후에 년, 월, 일로 분리하고 year를 인덱스로 셋팅하여 출력하세요.


In [196]:
df = pd.read_csv('./dataset/stock-data.csv')
df.head()

Date  Close  Start   High    Low  Volume
0  2018-07-02  10100  10850  10900  10000  137977
1  2018-06-29  10700  10550  10900   9990  170253
2  2018-06-28  10400  10900  10950  10150  155769
3  2018-06-27  10900  10800  11050  10500  133548
4  2018-06-26  10800  10900  11000  10700   63039

In [197]:
df['Date'] = pd.to_datetime(df['Date'])
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20 entries, 0 to 19
Data columns (total 6 columns):
 #   Column  Non-Null Count  Dtype         
---  ------  --------------  -----         
 0   Date    20 non-null     datetime64[ns]
 1   Close   20 non-null     int64         
 2   Start   20 non-null     int64         
 3   High    20 non-null     int64         
 4   Low     20 non-null     int64         
 5   Volume  20 non-null     int64         
dtypes: datetime64[ns](1), int64(5)
memory usage: 1.1 KB


In [201]:
df['year'] = df['Date'].dt.year
df['month'] = df['Date'].dt.month
df['day'] = df['Date'].dt.day
df.head()

Date  Close  Start   High    Low  Volume  year  month  day
0 2018-07-02  10100  10850  10900  10000  137977  2018      7    2
1 2018-06-29  10700  10550  10900   9990  170253  2018      6   29
2 2018-06-28  10400  10900  10950  10150  155769  2018      6   28
3 2018-06-27  10900  10800  11050  10500  133548  2018      6   27
4 2018-06-26  10800  10900  11000  10700   63039  2018      6   26

In [202]:
df['Date'].dt.year

0     2018
1     2018
2     2018
3     2018
4     2018
5     2018
6     2018
7     2018
8     2018
9     2018
10    2018
11    2018
12    2018
13    2018
14    2018
15    2018
16    2018
17    2018
18    2018
19    2018
Name: Date, dtype: int64

In [113]:
df.set_index('year',inplace = True)
df.head()

Date  Close  Start   High    Low  Volume  month  day
year                                                           
2018 2018-07-02  10100  10850  10900  10000  137977      7    2
2018 2018-06-29  10700  10550  10900   9990  170253      6   29
2018 2018-06-28  10400  10900  10950  10150  155769      6   28
2018 2018-06-27  10900  10800  11050  10500  133548      6   27
2018 2018-06-26  10800  10900  11000  10700   63039      6   26

### Q34. './dataset/stock-data.csv'를 데이터프레임으로 불러와서 datetime64 자료형으로 변환한 후에 Period로 변환하고 년월일 표기 변경하여 년월을 인데스로 셋팅하여 출력하세요.


In [24]:
df = pd.read_csv('./dataset/stock-data.csv')
df.head()

Date  Close  Start   High    Low  Volume
0  2018-07-02  10100  10850  10900  10000  137977
1  2018-06-29  10700  10550  10900   9990  170253
2  2018-06-28  10400  10900  10950  10150  155769
3  2018-06-27  10900  10800  11050  10500  133548
4  2018-06-26  10800  10900  11000  10700   63039

In [25]:
df['Period'] = pd.to_datetime(df['Date'])
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20 entries, 0 to 19
Data columns (total 7 columns):
 #   Column  Non-Null Count  Dtype         
---  ------  --------------  -----         
 0   Date    20 non-null     object        
 1   Close   20 non-null     int64         
 2   Start   20 non-null     int64         
 3   High    20 non-null     int64         
 4   Low     20 non-null     int64         
 5   Volume  20 non-null     int64         
 6   Period  20 non-null     datetime64[ns]
dtypes: datetime64[ns](1), int64(5), object(1)
memory usage: 1.2+ KB


In [26]:
df.drop("Date", axis = 1, inplace = True)

In [27]:
df.head()

Close  Start   High    Low  Volume     Period
0  10100  10850  10900  10000  137977 2018-07-02
1  10700  10550  10900   9990  170253 2018-06-29
2  10400  10900  10950  10150  155769 2018-06-28
3  10900  10800  11050  10500  133548 2018-06-27
4  10800  10900  11000  10700   63039 2018-06-26

In [28]:
#Period를 인덱스로 변환
df.set_index('Period', inplace = True)
df.head()

Close  Start   High    Low  Volume
Period                                        
2018-07-02  10100  10850  10900  10000  137977
2018-06-29  10700  10550  10900   9990  170253
2018-06-28  10400  10900  10950  10150  155769
2018-06-27  10900  10800  11050  10500  133548
2018-06-26  10800  10900  11000  10700   63039

In [29]:
# 인덱스로 변환 후에 to_period로 바꿔줘야 한다.
df2 = df.to_period(freq = "M")
df2.head()

Close  Start   High    Low  Volume
Period                                     
2018-07  10100  10850  10900  10000  137977
2018-06  10700  10550  10900   9990  170253
2018-06  10400  10900  10950  10150  155769
2018-06  10900  10800  11050  10500  133548
2018-06  10800  10900  11000  10700   63039

In [33]:
df2 = df.to_period(freq = "Y")
df2.head()

Close  Start   High    Low  Volume
Period                                    
2018    10100  10850  10900  10000  137977
2018    10700  10550  10900   9990  170253
2018    10400  10900  10950  10150  155769
2018    10900  10800  11050  10500  133548
2018    10800  10900  11000  10700   63039